Implement node Id interface

cache-nez opened the pull request.

cache-nez committed:
Implement Id

This commit introduces an array-based implementation of Id: an ArrayId is
constructed from a mino.Address by taking a hash of the address
modulo (base ^ len), then presenting the id as an array of digits.
GetLength(), GetBase() and GetDigit(pos) operations are then trivial.

cache-nez committed:
Implement a Prefix of an Id

cache-nez committed:
Fix computing bigint representation of a hash

gnarula commented on id/mod.go:
```diff
@@ -1,7 +1,60 @@
+
 type Id interface {
-	GetLength() byte
+	GetLength() int
```
Consider removing the `Get` prefix here. It's idiomatic in Go to have a getter without the `Get` prefix in method names.

cache-nez replied:
Done, thank you! Does this also apply to `GetDigit`, which is not a simple getter?

gnarula commented on id/mod.go:
```diff
@@ -1,7 +1,60 @@
\ No newline at end of file
+}
+
+type ArrayId struct {
```
`ArrayId` => `ArrayNodeID`

gnarula commented on id/mod.go:
```diff
@@ -1,7 +1,60 @@
+	"math/big"
+)
+
 type Id interface {
```
`Id` to `NodeID`

gnarula commented on id/mod.go:
```diff
@@ -1,7 +1,60 @@
+}
+
+type ArrayId struct {
+	Id   []byte
```
nit, `ID` here

cache-nez replied:
Left it as a private member for now. Do you think it makes sense to make it public?

gnarula commented on id/mod.go:
```diff
@@ -1,7 +1,60 @@
+	return id.Id[pos]
+}
+
+func hash(addr mino.Address) (h *big.Int) {
```
Replace this with a non named return please, i.e. have a `return h` at the end

gnarula commented on id/mod.go:
```diff
@@ -1,7 +1,60 @@
+	return ArrayId{id, base}
+}
+
+func (id ArrayId) GetLength() int {
```
Please add comments for all public methods

gnarula commented on id/mod.go:
```diff
@@ -1,7 +1,60 @@
+}
+
+// Constructs an Id by taking a hash of the address modulo (base ^ len), then presenting the id as an array of digits.
+func MakeArrayId(address mino.Address, base byte, len int) Id {
```
A more idomatic name would be `NewArrayNodeID` with a return type of `ArrayNodeID`. Please ensure the comment wraps around 80 chars.

Would be useful to have a unit test for this

gnarula commented on id/prefix.go:
```diff
@@ -0,0 +1,33 @@
+package id
+
+type Prefix interface {
+	GetLength() int
```
Nit, consider removing `Get`

gnarula commented on id/prefix.go:
```diff
@@ -0,0 +1,33 @@
+
+type Prefix interface {
+	GetLength() int
+	isPrefixOf(Id) bool
```
Is there a reason this method is private?

cache-nez replied:
No, thank you for the catch!

gnarula commented on id/prefix.go:
```diff
@@ -0,0 +1,33 @@
+}
+
+// Digits of the prefix have to be stored in a string, because this structure is used as a key in the routing table map.
+type PrefixImpl struct {
```
Perhaps `StringPrefix` might be a better name

gnarula reviewed (changes requested):
Added some comments about Go conventions in general. Please consider having a unit test for `MakeArrayId`. Looks good otherwise.

cache-nez committed:
Fix computation of ArrayId: the base of id was unintentionally set to base ^ len

cache-nez committed:
Style changes to ids: better names, refactor hash() into two functions, comment all methods

cache-nez committed:
Add Equals() method to NodeID interface

cache-nez committed:
When possible, use the hash of mino.Address:Marshal instead of String() to construct an id

cache-nez committed:
Ignore if the node is an orchestrator when constructing its id.

This is a hack to accommodate for minogrpc's design:
1) the first byte is used to indicate if a node is orchestrator or not
2) the only way to reach the orchestrator is to route a message to nil
   from its server side, which has the same address but orchestrator byte
   set to f.
We therefore have to ignore if a node is the orchestrator to be able to
route the message first to its server side, then from the server side to
the client side.

cache-nez committed:
Add tests for ArrayNodeID

cache-nez committed:
Implement and test computing common prefix of two ids

gnarula commented:
Thanks for adding the tests.One small comment following Noemien's remark in the other PR. Comments in Go begin with the method name. Ready to be merged after that :)

cache-nez committed:
Add documenting comments to methods
